### DISCUSSION TITLE
Dispatch actions inside ICommand derived class to use MVVM

### DISCUSSION THREAD

**Discussion by Shuusui** [mrpmorris/Fluxor#411] [Q&A]
I'm fairly new to Fluxor so far and don't really have that much experience with the concept either, so it might be a relative obvious solution to my problem or maybe no solution because Fluxor and MVVM contradicts each other. 
I currently use Blazorise as library to be able to bind ICommand derived properties to my button executions etc. I have a button to submit my login form to the server and receive an authentication token out of it. What I would like to do now is after I received the authentication token from the server to launch an action to change the state to "logged in" so that my View can update, i.e. remove the Login/Register components and add components to get used when the user is logged in. 
I know how I can do that theoretically like I could probably pass the dispatcher to my command or don't do it in a command but I wonder if there would be a more.... convenient solution for what I'm trying or if maybe Fluxor is not the right library to use when I also want to do MVVM. 
I also thought it would maybe make sense to run the dispatcher when I click on my button right away but it doesn't really seem right since I would first need to send data to the server, receive the token and only if I actually receive a token I'd like to update the state. I could probably update the action anyway even if I don't get an authentication token back but it still doesn't look like the right approach to me.

**Comment by mrpmorris** [comment]
1. Dispatch an action
2. Effect is triggered by this action, which goes to the server for the info you need
3. Once returned, the Effect triggers another action is dispatched indicating if it was a success/failure

Reducers update the state as appropriate for each action.

**Reply by Shuusui** [reply]
Thank you for your answer but it would still not make it possible for me to dispatch the action in my ICommand implementing class. I would then dispatch the action in my component which is not quite what I would like to do. 
What I did now was just passing the dispatcher to my command which is bound to my button but this doesn't look very clean. It would be very nice if I could keep my MVVM style command binding but also use Fluxor actions within does bound commands.

**Reply by mrpmorris** [reply]
Can you inject the IDispatcher into your class that deals with the ICommand?

I am not familiar enough with MVVM, perhaps you can show me some code?

**Reply by Shuusui** [reply]
Yes kind of. My component has the viewmodel which itself has the model (from MVVM) but either the viewmodel or the model then has the ICommand, so I would need to pass the dispatcher to the one with the command so that it can pass the dispatcher to the actual ICommand. So I would just pass the dispatcher around since I don't know if I can or could launch actions without it somehow. That's what I meant with it's not really clean to use it like that.

**Reply by mrpmorris** [reply]
Passing dependencies like this isn't a problem, but I think it would be better to either have the creator of the command dispatch it instead, or whatever receives the command should dispatch it.

But I'm working blind here.
